teps/0094-configuring-resources-at-runtime.md

```diff
@@ -52,12 +61,495 @@ While this may be a valuable feature, it should be considered in a separate prop
+   // Any field present in both a Sidecar and its ContainerOverride
+   // will use the value from the ContainerOverride.
+   SidecarOverrides map[string]ContainerOverride
+}
+
+type PipelineTaskRunSpec struct {
+  ...
+   // A map of Step names to ContainerOverrides to apply to that Step.
+   // A single entry with key "*" will apply the ContainerOverride to all Steps.
+   // Any field present in both a Step and its ContainerOverride
+   // will use the value from the ContainerOverride.
+   StepOverrides map[string]ContainerOverride
+
+   // A map of Sidecar names to ContainerOverrides to apply to that Sidecar.
+   // A single entry with key "*" will apply the ContainerOverride to all Sidecars.
+   // Any field present in both a Sidecar and its ContainerOverride
+   // will use the value from the ContainerOverride.
+   SidecarOverrides map[string]ContainerOverride
+}
+```
+
+### Risks and Mitigations
+
+This proposal may increase confusion for users. It may not be obvious which of 
+`Task.Step.Resources`, `Task.StepTemplate.Resources`, and `TaskRun.StepResources`
+takes precedence (see [Merging resource requirements](#merging-resource-requirements)).
+In addition, `StepResources` and `SidecarResources` may be confused with `PipelineResource`s.
+This risk can be mitigated via documentation.
+
+### User Experience
+
+The current workaround for lack of this feature is to write a new `Task`
+for each set of resource constraints, as described in
+[this comment](https://github.com/tektoncd/pipeline/issues/4080#issuecomment-884958486)
+from a buildah user.
+
+This proposal moves environment-specific configuration into `TaskRun` definitions,
+allowing `Task` definitions to be reused.
+We may want to add an option to the CLI to specify resource requirements when starting
+a `TaskRun` via `tkn task start`, but this is not necessary for the initial implementation.
+
+## Test Plan
+
+Unit tests should suffice for this feature, covering the following cases:
+- `Task`s with no resource requirements specified
+- `Task`s with resource requirements that are partially overridden by the `TaskRun`
+- `Task`s with resource requirements that are fully overridden by the `TaskRun`
+- `TaskRun`s with resource requirements launched by themselves and from `PipelineRun`s.
+
+Examples should be included for overriding resource requirements
+(e.g. the example used in [Merging resource requirements](#merging-resource-requirements)).
+
+## Design Evaluation
+### Reusability
+This proposal increases reusability of `Task`s and `Pipeline`s by allowing
+environment-specific execution requirements to be updated at runtime.
+
+### Simplicity
+The proposed solution contains the minimum number of features that meet the specified
+requirements, compared to the listed alternatives. 
+
+### Flexibility
+This proposal increases Tekton's flexibility by giving users more options to modify `Task`s.
+There isn't a clear strategy for implementing this functionality via a plugin system.
```

Comment:
one con for flexibility: imo introducing the `*` syntax could be the beginning of implementing our own expression syntax

Reply:
I agree, I've removed this part of the proposal.